core/modules/utils/linked-list.js

```diff
@@ -0,0 +1,106 @@
+/*\
+module-type: utils
+title: $:/core/modules/utils/linkedlist.js
+type: application/javascript
+
+This is a doubly-linked indexed list intended for manipulation, particularly
+pushTop, which it does with significantly better performance than an array.
+
+\*/
+(function(){
+
+function LinkedList() {
+	this.clear();
+};
+
+LinkedList.prototype.clear = function() {
+	this.index = Object.create(null);
+	this.next = this;
+	this.prev = this;
+	this.length = 0;
+};
+
+LinkedList.prototype.remove = function(value) {
+	if ($tw.utils.isArray(value)) {
```

Comment:
in terms of syntax/code style @flibbles I believe we do not use spaces between the `if` and the `(`

Reply:
Ooh, right. I knew there was a javascript convention for TW somewhere. Couldn't find it though, so I guessed. I'll fix this.

Reply:
That's correct